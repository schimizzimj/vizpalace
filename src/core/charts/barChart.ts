import * as d3 from "d3";
import {
  initializeChartDimensions,
  createSVG,
  appendXAxis,
  appendYAxis,
} from "../base/chartUtils";
import { BarData, ChartOptions } from "../../types";

export function createBarChart(
  element: HTMLElement,
  data: BarData[],
  userOptions: Partial<ChartOptions> = {}
): void {
  // Merge default options with user options
  const options = initializeChartDimensions(userOptions);
  let { width, height, animationDuration } = options;
  const svg = createSVG(element, options);
  animationDuration = animationDuration ?? 1000;

  const xScale = d3
    .scaleBand()
    .domain(data.map((d) => d.label))
    .range([0, width])
    .padding(0.1);

  const yMax =
    d3.max(data, (d) =>
      d.value !== undefined && d.value !== null ? d.value : 0
    ) || 0;

  const yScale = d3.scaleLinear().domain([0, yMax]).range([height, 0]);

  const bars = svg
    .selectAll(".bar")
    .data(data)
    .enter()
    .append("rect")
    .attr("class", "bar")
    .attr("x", (d) => xScale(d.label) || 0)
    .attr("y", height)
    .attr("width", xScale.bandwidth())
    .attr("height", 0);

  bars
    .transition()
    .duration(animationDuration)
    .ease(d3.easeCubicInOut)
    .attr("y", (d) => yScale(d.value ?? 0))
    .attr("height", (d) => height - yScale(d.value ?? 0));

  appendXAxis(svg, xScale, height);
  appendYAxis(svg, yScale);
}
