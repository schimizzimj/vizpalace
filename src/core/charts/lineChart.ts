import * as d3 from "d3";
import {
  initializeChartDimensions,
  createSVG,
  appendXAxis,
  appendYAxis,
} from "../base/chartUtils";

export function createLineChart(
  element: HTMLElement,
  data: LineData[],
  userOptions: Partial<ChartOptions> = {}
): void {
  const options = initializeChartDimensions(userOptions);
  let { width, height, animationDuration } = options;

  const svg = createSVG(element, options);
  animationDuration = animationDuration ?? 0;

  const xScale = d3
    .scaleTime()
    .domain(d3.extent(data, (d) => d.date) as [Date, Date])
    .range([0, width]);

  const yMax =
    d3.max(data, (d) =>
      d.value !== undefined && d.value !== null ? d.value : 0
    ) || 0;

  const yScale = d3.scaleLinear().domain([0, yMax]).range([height, 0]);

  // Create the line generator
  const line = d3
    .line<LineData>()
    .x((d) => xScale(d.date))
    .y((d) => yScale(d.value ?? 0));

  // Draw the line
  const linePath = svg
    .append("path")
    .datum(data)
    .attr("class", "line")
    .attr("d", line)
    .attr("fill", "none")
    .attr("stroke", "steelblue")
    .attr("stroke-width", 1.5);

  if (linePath.node()?.getTotalLength) {
    const totalLength = linePath.node()?.getTotalLength() ?? 0;
    linePath
      .attr("stroke-dasharray", `${totalLength} ${totalLength}`)
      .attr("stroke-dashoffset", totalLength)
      .transition()
      .duration(animationDuration)
      .ease(d3.easeLinear)
      .attr("stroke-dashoffset", 0);
  }

  appendXAxis(svg, xScale, height);
  appendYAxis(svg, yScale);
}
