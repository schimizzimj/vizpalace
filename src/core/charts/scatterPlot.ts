import * as d3 from "d3";
import { ChartData, ScatterPlotOptions } from "../../types";
import { createSVG, initializeScatterPlotDimensions } from "../base/chartUtils";
import { appendXAxis, appendYAxis } from "../base/axes";
import { isScaleBand } from "../utils/utils";

export function createScatterPlot<T = number | string | Date>(
  element: HTMLElement,
  data: ChartData<T>,
  userOptions: Partial<ScatterPlotOptions> = {}
) {
  const options = initializeScatterPlotDimensions(userOptions);
  let { chartDimensions, animationDuration, xAxis, yAxis } = options;
  const { width, height } = chartDimensions;
  const svg = createSVG(element, options);
  const duration = animationDuration ?? 0;

  const allDataPoints = data.flatMap((series) => series.values);

  let xScale:
    | d3.ScaleBand<string>
    | d3.ScaleTime<number, number>
    | d3.ScaleLinear<number, number>;

  if (typeof allDataPoints[0].x === "string") {
    xScale = d3
      .scaleBand()
      .padding(0.1)
      .domain(allDataPoints.map((d) => d.x as string));
  } else if (allDataPoints[0].x instanceof Date) {
    xScale = d3
      .scaleTime()
      .domain(d3.extent(allDataPoints, (d) => d.x as Date) as [Date, Date]);
  } else {
    xScale = d3
      .scaleLinear()
      .domain(
        d3.extent(allDataPoints, (d) => d.x as number) as [number, number]
      );
  }

  xScale.range([0, width]);

  const yMin = d3.min(allDataPoints, (d) => d.y ?? 0) ?? 0;
  const yMax = d3.max(allDataPoints, (d) => d.y ?? 0) ?? 0;

  const yScale = d3.scaleLinear().domain([yMin, yMax]).range([height, 0]);

  let circlesWrapper: d3.Selection<SVGGElement, any, any, any> =
    svg.select(".circles-wrapper");

  if (circlesWrapper.empty()) {
    circlesWrapper = svg.append("g").attr("class", "circles-wrapper");
  }

  const circles = circlesWrapper
    .selectAll<SVGCircleElement, any>(".circle")
    .data(allDataPoints);

  circles.exit().remove();

  circles
    .enter()
    .append("circle")
    .attr("class", "circle")
    .attr("cx", (d) => {
      if (isScaleBand(xScale)) {
        return (xScale(d.x as string) ?? 0) + xScale.bandwidth() / 2;
      } else {
        return xScale(d.x as any) ?? 0;
      }
    })
    .attr("cy", (d) => yScale(d.y ?? 0) ?? 0)
    .attr("r", 5)
    .attr("fill", "steelblue")
    .merge(circles)
    .transition()
    .duration(duration)
    .attr("cx", (d) => {
      if (isScaleBand(xScale)) {
        return (xScale(d.x as string) ?? 0) + xScale.bandwidth() / 2;
      } else {
        return xScale(d.x as any) ?? 0;
      }
    })
    .attr("cy", (d) => yScale(d.y ?? 0) ?? 0);

  appendXAxis(svg, xScale, chartDimensions, xAxis ?? {});
  appendYAxis(svg, yScale, chartDimensions, yAxis ?? {});
}
