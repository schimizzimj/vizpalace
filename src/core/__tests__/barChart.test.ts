import { assert, expect, test, describe, beforeEach, afterEach } from "vitest";
import { createBarChart } from "../charts/barChart";

describe("barChart", () => {
  let container: HTMLDivElement;
  let data: BarData[];

  beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
    data = [
      { label: "foo", value: 10 },
      { label: "bar", value: 20 },
      { label: "baz", value: 30 },
    ];
  });

  afterEach(() => {
    document.body.removeChild(container);
  });

  test("it should render a bar chart", () => {
    createBarChart(container, data, {});

    const svg = container.querySelector("svg");
    assert(svg);

    const bars = svg.querySelectorAll("rect");
    expect(bars.length).toBe(3);
  });
});
