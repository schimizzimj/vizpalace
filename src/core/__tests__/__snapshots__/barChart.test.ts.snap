// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`barChart > it should look visually correct 1`] = `"<svg width=\\"540\\" height=\\"350\\"><rect class=\\"bar\\" x=\\"17.419354838709694\\" y=\\"233.33333333333337\\" width=\\"156.7741935483871\\" height=\\"116.66666666666663\\"></rect><rect class=\\"bar\\" x=\\"191.61290322580646\\" y=\\"116.66666666666669\\" width=\\"156.7741935483871\\" height=\\"233.33333333333331\\"></rect><rect class=\\"bar\\" x=\\"365.80645161290323\\" y=\\"0\\" width=\\"156.7741935483871\\" height=\\"350\\"></rect></svg>"`;
